backend/core/rule/tree.py

```diff
@@ -10,31 +10,65 @@
+    def __init__(self, rule_name, sugang_list, metadata, course_model):
+        base_dir = os.path.dirname(os.path.abspath(__file__))
+        f = open(os.path.join(base_dir, 'rules', rule_name + '.yml'), 'r', encoding='utf-8')
         self.tree = yaml.load(f)
         f.close()
         self.sugang_list = sugang_list
         self.metadata = metadata
-        self.base_node = TreeNode(None, and_func(), '!GRADUATE', self.metadata)
+        self.course_model = course_model
+        self.base_node = TreeNode(None, and_func(), '!GRADUATE', self.metadata, course_model,
+                                  [0, 0, False], False, False)
         self.load_tree(self.tree, self.base_node)
 
     def load_tree(self, current_tree, previous_node):
         for course in current_tree:
             if isinstance(course, str):
-                previous_node.add_children(TreeNode(course, None, self.metadata, None, self.sugang_list))
+                if course.startswith('$'):
+                    subarea = course[1:]
+                    subarea_code_list = [item.code for item in self.course_model.objects.filter(subarea=subarea)]
+                    code_set = set(subarea_code_list)
+                    for code in code_set:
+                        previous_node.add_children(
+                            TreeNode(code, None, None, self.metadata, self.course_model,
+                                     [0, 0, False], False, False, self.sugang_list))
+                else:
+                    previous_node.add_children(
+                        TreeNode(course, None, None, self.metadata, self.course_model,
+                                 [0, 0, False], False, False, self.sugang_list))
             elif type(course) is dict:
                 args = course.get('args', [])
+                hide_false = course.get('hide_false', False)
+                required_credit = course.get('required_credit', 0)
+                sum_false = course.get('sum_false', False)
+                delete_used = course.get('delete_used', False)
                 current_func = course['func']
                 if current_func == 'if':
                     if not if_func(*args, **self.metadata):
                         continue
                     self.load_tree(course['then'], previous_node)
+                elif current_func == 'if_not':
+                    if not if_not_func(*args, **self.metadata):
+                        continue
+                    self.load_tree(course['then'], previous_node)
+                elif current_func == 'if_greater':
+                    if not if_greater_func(*args, **self.metadata):
+                        continue
+                    self.load_tree(course['then'], previous_node)
+                elif current_func == 'if_less':
+                    if not if_less_func(*args, **self.metadata):
+                        continue
+                    self.load_tree(course['then'], previous_node)
                 else:
```

Comment:
These codes are redundant, and would be better if they're generalized.
We can use 'dictionary of functions', or maybe '[getattr()](http://stackoverflow.com/questions/9168340/python-using-a-dictionary-to-select-function-to-execute)' with some thinking.